ambry-commons/src/main/java/com.github.ambry.commons/ResponseHandler.java

```diff
@@ -114,4 +116,19 @@ public void onEvent(ReplicaId replicaId, Object event) {
       onNetworkEvent(replicaId, (NetworkClientErrorCode) event);
     }
   }
+
+  /**
+   * Take action when connection to certain {@link DataNodeId} timed out. The clustermap is supposed to mark node resource
+   * down and avoid subsequent requests routed to this node within down window.
+   * @param dataNodeId the {@link DataNodeId} associated with timeout connection.
+   */
+  public void onConnectionTimeout(DataNodeId dataNodeId) {
+    if (dataNodeId != null) {
```

Comment:
Usually it shouldn't be `null`, I added this check if `dataNodeId` is not found, for some reason, for the timeout connection. In this case, we ignore the timeout connection instead of NPE here.

Reply:
Is there any known case that causes null? Or this is just a guard?

Reply:
No in this PR but `dataNodeId` is parsed from ResponseInfo where we add datanode info in NetworkClient. I think it's good to add a check here if we make changes or add tests in the NetworkClient without specifying datanode in ResponseInfo.